iodata/iodata.py

```diff
@@ -252,31 +253,45 @@ class IOData:
     @property
     def atcorenums(self) -> np.ndarray:
         """Return effective core charges."""
-        result = self._atcorenums
-        if result is None and self.atnums is not None:
+        if self._atcorenums is None and self.atnums is not None:
             # Known bug in pylint. See
             # https://stackoverflow.com/questions/47972143/using-attr-with-pylint
             # https://github.com/PyCQA/pylint/issues/1694
             # pylint: disable=no-member
-            result = self.atnums.astype(float)
-            self._atcorenums = result
-        return result
+            self.atcorenums = self.atnums.astype(float)
+        return self._atcorenums
 
     @atcorenums.setter
     def atcorenums(self, atcorenums):
+        if atcorenums is None:
+            if self.nelec is not None:
+                # Set _charge because charge can no longer be derived from
+                # atcorenums and nelec.
+                self._charge = self._atcorenums.sum() - self.nelec
```

Comment:
I think this line will raise an error if `nelec` was set before `atcorenums` and the latter is set to None afterwards.

Reply:
Thanks. I'll add a unit test for this scenario, and make sure it works.

Reply:
You were right! It should be fixed now. I also found another minor issue with the dtype of atcorenums, which is fixed too.